File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def fc(diffs, times, limit, m):
    # 이분 탐색: 고른 mid 값이 유효하면 true, 아니면 false
    totalTime = 0
    time_prev = 0
    n = len(diffs)
    
    for i in range(n):
        gap = m - diffs[i]
        time_cur = times[i]
        if (gap >= 0):
            totalTime += time_cur
        else:
            totalTime += (-gap) * (time_prev + time_cur) + time_cur
        time_prev = time_cur
    
    # print(totalTime)
    if (totalTime <= limit):
        return True
    else:
        return False
            

def solution(diffs, times, limit):
    
    
    
    l = 0
    r = 10 ** 15 + 1
    # mid = (l + r) // 2
    # fc(diffs, times, limit, mid)
    
    while (l + 1 < r):
        mid = (l + r) // 2
        
        # level을 낮춰도 괜찮음
        if (fc(diffs, times, limit, mid)):
            r = mid
            
        # 해당 level 보다는 높아야함
        else:
            l = mid
    
    return r

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;
typedef long long ll;

bool fc(vector<int>& diffs, vector<int>& times, ll limit, ll mid){
    ll total = 0;
    ll time_prev = 0;
    
    for (int i = 0; i < diffs.size(); i++) {
        ll time_cur = times[i];
        
        if (mid - diffs[i] >= 0){
            total += time_cur;
        } else {
            total += (time_cur + time_prev) * (diffs[i] - mid) + time_cur;
        }
        time_prev = time_cur;
        
    }
    
    // cout << total << '\n';
    if (total <= limit) {
        return true;
    } else {
        return false;
    }
}

int solution(vector<int> diffs, vector<int> times, ll limit) {
    ll l = 0;
    ll r = 1000000000000000 + 1;
    
    while (l + 1 < r){
        ll mid = (l + r) / 2;
        
        // 레벨을 낮춰도 됨
        if (fc(diffs, times, limit, mid)){
            r = mid;
        } else {
            // 해당 레벨보다는 높아야함.
            l = mid;
        }
    }
    
    return r;
}